### PR TITLE
[ENH] First transfer of deep learning classifiers and regressors from sktime-dl

### PR DESCRIPTION
**Pull request by TonyBagnall**
this is my first go at this and open to discussion as to structure. Once we have agreed how to do CNN, the rest of the algorithms will come over quickly. However, there are features here some may dislike, so want to out there early. Points to note
1. Introduces a whole new module called networks. This contains the network structures that are shared by both classifiers and regressors. This is necessary because it was agreed long ago to split the algorithms into regressors/classifiers (rather than have one class with both capabilities, which was my preferred structure)
2. classifiers go in classification/deep_learning, and there is an abstract class there that inherits from BaseClassifier. This intermediate class is there because we can define default DL classifier predict_proba behaviour, and we want an extra abstract method.

In terms of the implementation, this is not yet correct for CNN, and there is not yet any testing. Following issues

1. Input X: sktime uses (n,d,m) but it seems keras assumes (n,m,d), easy to fix but I need to confirm this, but its a bit annyoing
2. Input Y: I'm hoping we dont need the label encoding, and can just use the one hot encoder if required, again needs resolving with base class. The issue here is whether keras assumes classes are mapped onto 0,...,n_classes-1 or not. 
3. The validation data. This is trickier, since this is redefining _fit from the base, which is undesirable. I would like input/opinions on preferred way of setting validation data in training. 
4. I need to sort out the random state, but that should be easy
5. Need to check it works with soft dependency set up

### CODE REVIEW

File: sktime/classification/base.py [hunk 1]
```diff
@@ -42,12 +41,12 @@ class BaseClassifier(BaseEstimator, ABC):
...
 
     The base classifier specifies the methods and method signatures that all
-    classifiers have to implement. Attributes with a underscore suffix are set in the
+    classifiers have to implement. Attributes with an underscore suffix are set in the
     method fit.
 
     Parameters
     ----------
-    classes_            : ndarray of class labels, possibly strings
+    classes_            : array of class labels, possibly strings
```
Comment (fkiraly) [inline sktime/classification/base.py:49]: should be `np.ndarray` if that's the class/type, `array` is not a class/type

File: sktime/classification/deep_learning/base.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            The shape of the data fed into the input layer
+        n_classes: int
+            The number of classes, which shall become the size of the output
+            layer
+
+        Returns
+        -------
+        output : a compiled Keras Model
+        """
+        ...
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base.py:64]: is sth missing here?

File: sktime/classification/deep_learning/base.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    def _predict(self, X, **kwargs):
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base.py:66]: I think we have no `kwargs` in the functions? Or do we want them?

File: sktime/classification/deep_learning/base.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        probs = self.model_.predict(X, self.batch_size, **kwargs)
+
+        # check if binary classification
+        if probs.shape[1] == 1:
+            # first column is probability of class 0 and second is of class 1
+            probs = np.hstack([1 - probs, probs])
+
+        return probs
+
+    def convert_y(self, y, label_encoder=None, onehot_encoder=None):
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base.py:100]: maybe call this `convert_y_to_keras` or similar?
to avoid confusion and accidental overrides

File: sktime/classification/deep_learning/cnn.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    def __init__(
+        self,
+        n_epochs=2000,
+        batch_size=16,
+        kernel_size=7,
+        avg_pool_size=3,
+        n_conv_layers=2,
+        callbacks=None,
+        verbose=False,
+    ):
```
Comment (fkiraly) [inline sktime/classification/deep_learning/cnn.py:60]: there should be a `_check_soft_dependencies` here too, at error severity level

File: sktime/classification/deep_learning/cnn.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        """
+        if self.callbacks is None:
+            self._callbacks = []
+        y_onehot = self.convert_y(y)
+        check_random_state(self.random_state)
+
+        # Transpose to conform to Keras input style.
+        X = X.transpose(0, 2, 1)
+        self.input_shape = X.shape[1:]
+        self.model_ = self.build_model(self.input_shape, self.n_classes_)
```
Comment (fkiraly) [inline sktime/classification/deep_learning/cnn.py:132]: this looks like much of it will be the same - should this be boilerplate that is in the intermediate base class?
Can you perhaps get rid of `_fit` entirely?

**Review by fkiraly — COMMENTED** [review]
had a quick look - I think, high-level, code structure makes sense.

Having the "intermediate base class" or "adapter base class" is the right idea - I'd encourage to think about how as much as possible keras boilerplate can go there. I feel some of what is in `_fit` and perhaps `build_model` could move there.

**Comment by TonyBagnall** [comment]
> had a quick look - I think, high-level, code structure makes sense.
> 
> Having the "intermediate base class" or "adapter base class" is the right idea - I'd encourage to think about how as much as possible keras boilerplate can go there. I feel some of what is in `_fit` and perhaps `build_model` could move there.

I agree, but I think I want to wait until I do the subsequent ones until I see how much can be moved up to the parent class, the commonality will be clearer then

**Comment by fkiraly** [comment]
> I agree, but I think I want to wait until I do the subsequent ones until I see how much can be moved up to the parent class, the commonality will be clearer then

Sure - I assume you want to refactor then as you develop, before this PR is final?

File: sktime/classification/deep_learning/base.py (hunk 2 above)
Comment (TonyBagnall) [reply sktime/classification/deep_learning/base.py:64]: I've removed the text output : it returns a keras Model object,

File: sktime/classification/deep_learning/base.py (hunk 3 above)
Comment (TonyBagnall) [reply sktime/classification/deep_learning/base.py:66]: dont need kwargs, I've removed them, not used

File: sktime/classification/deep_learning/cnn.py (hunk 5 above)
Comment (TonyBagnall) [reply sktime/classification/deep_learning/cnn.py:60]: I've raised the severity to error, but was then wondering whether I could I not just do it in the base classes rather than in every individual file, I can try it out with the second one

File: sktime/classification/deep_learning/cnn.py (hunk 6 above)
Comment (TonyBagnall) [reply sktime/classification/deep_learning/cnn.py:132]: possibly, I'll look through on the next PR to see if it could be moved up to the base.

**Comment by TonyBagnall** [comment]
I'd like to complete the port over several PRs to avoid one massive PR that drags on for months. I think this can go in as the first of several, I will do the regression equivalent next with a refactor of the regressor base class to fit/_fit etc

**Review by fkiraly — APPROVED** [review]
OK, makes sense.

Some remarks, blocking until resolved or discussed:

* The parameters of `CNNNetwork` cannot be accessed from `CNNClassifier` via the `sklearn` interface. Should this not be a component, i.e., a constructor arg? And should `BaseDeepNetwork` not inherit from `BaseObject` so these parameters can be accessed?
* a name `CNNClassifier` is pretty generic. There is not going to be only a single CNN classifier, is there? Perhaps a better name would be appropriate?
* the entire conversion business at the start of `_fit` looks *again* like conversion boilerplate that you intend to multiply across all classes inheriting from `BaseDeepClassifier`. This looks like a prime case for an mtype that applies to the `y`. You know, `y_inner_mtype = "keras_labels"` or similar.

**Review by fkiraly — CHANGES_REQUESTED** [review]
mis-clicked, the above should have been "change request"

**Comment by fkiraly** [comment]
> so you mean containment rather than inheritance? If not, could you mock me an example of what you mean? I'm not sure that I really want to build an interface between sklearn and keras, I'm sure such things already exist, but this is is just the first vanilla port from sktime-dl to sktime. Happy for BaseDeepNetwork to inherit from our BaseObject,

Would be a simple change imo (may be too naive though). Idea is, make the following two changes:

1. let `BaseDeepNetwork` inherit from `BaseObject` (I see you have already done this)

2. change `CNNClassifier` as follows:

```python

from sklearn.base import clone

class CNNClassifier(BaseDeepClassifier):

    def __init__(self, cnn_network, n_epochs=2000, etcetera)
        self.cnn_network = self.cnn_network
        self.cnn_network_ = clone(self.cnn_network)   # use this clone internally

    ...
line 86:
        input_layer, output_layer = self.cnn_network_.build_network(input_shape, **kwargs)
```

**Comment by fkiraly** [comment]
> I agree it can all move up to the base classifier if appropriate, but I need to check it is the same across all 13 classifiers before doing so.

Ok, I'm happy to do it as a refactor and not up-front if it is not clear that it is this precise boilerplate that will be used 13 times later. I consider this point addressed by your comment.

**Comment by fkiraly** [comment]
> this is a "vanilla" CNNNetwork, used in the most basic way, and it was called just CNN in Fawaz et al., whose naming conventions we have adopted.

Ok, fair - if it is common terminology in the field of time series classification, happy to keep it this way. Point addressed.

**Comment by TonyBagnall** [comment]
so to be clear, as the base model for all regressors and classifiers, you prefer containment to inheritance? I'm ok with that, but I want confirmation because I could have sworn you said previously that you liked it as it is.  Also I dont get this

**Comment by fkiraly** [comment]
> so to be clear, as the base model for all regressors and classifiers, you prefer containment to inheritance?

Can you clarify this, please? I do not understand your question, sorry.

I thought the base model for all classifiers was `BaseClassifier`, and the base model for all neural network classifiers was `BaseDeepClassifier`, neither of which would be impacted by this?

>did you mean `self.cnn_network = cnn_network`

yes, of course - silly mistake, sorry.

> Will `sklearn.clone` work with something that is not an sktime estimator?

You mean sklearn estimator, right?

No, it won't work with sth that's not an sklearn estimator, but `BaseObject` *is* an sklearn estimator!

**Comment by fkiraly** [comment]
> you prefer containment to inheritance?

Generally, yes,
https://en.wikipedia.org/wiki/Composition_over_inheritance
and especially if it looks like "putting a model/algorithm together from multiple pieces" (which the above looks like to me, but I'd appreciate if you'd help me understand the above point better).

**Comment by TonyBagnall** [comment]
> > so to be clear, as the base model for all regressors and classifiers, you prefer containment to inheritance?
> 
> Can you clarify this, please? I do not understand your question, sorry.

currently, sktime-dl uses multiple inheritance
image

I asked, probably on slack now I think about it, for expressions of preference as to inheritance vs composition/aggregation. I'll see if containment raises any issues, but I'm not sure it sure be configurable via the init arguments, the CNNClassifier should always use a CNNNetwork, and the parameters of that network should be parameters of the constructor to CNNClassifier, not the network itself. I think there should be an argument for activation="sigmoid",  loss="mean_squared_error", and metrics=["accuracy"] in the constructor though

**Comment by TonyBagnall** [comment]
> 
first test for inheritance vs containment is whether the relationship is "IS-A" or "HAS-A". A CNNclassifier could be considered to both be a network and have a network. In this situation, I would always prefer containment, specially when multiple inheritance is involved. However, my other null position is if it aint broke, dont fix it, especially as my understanding of keras is minimal. Given it was already implemented with multiple inheritance, I started off with that position. But given your preference, I'll try switch it to containment, and if it doesnt raise any problems with keras, proceed with that

**Comment by TonyBagnall** [comment]
switched to containment, seems ok, although there is some conflict between strings and keras, so its still hard coded as sigmoid and precommit has weirdly reformatted the input strings 
image

I also need to look at seeding, although this will always be a problem, I'm not sure you can make GPU deterministic, I'll look. I think this PR is a way from going in, I want to reproduce the published results and get it running on the GPU server before putting it in

**Comment by TonyBagnall** [comment]
weird this, it works with activation = "sigmoid" but not this 
image

**Comment by fkiraly** [comment]
>  the CNNClassifier should always use a CNNNetwork, and the parameters of that network should be parameters of the constructor to CNNClassifier

Well, but you could come up with another architecture of the network, no? So it would not always be a CNNNetwork, but sth that shares interface with it.

**Comment by fkiraly** [comment]
> weird this, it works with activation = "sigmoid" but not this

What do you mean by that?

**Comment by fkiraly** [comment]
you also want to remove the tuple brackets from around `loss` and `metric`, this was probably added by precommit after you left a comma there by accident

**Comment by TonyBagnall** [comment]
I have recreated the CNN results to my satisfaction from here, https://raw.githubusercontent.com/hfawaz/dl-4-tsc/master/results/results-ucr-128.csv so this is ready to go in as the first version. I will then take a deeper look at the rest to see if we can generalise more code into the base

File: sktime/networks/cnn.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+"""Time Convolutional Neural Network (CNN) (minus the final output layer)."""
+
+__author__ = ["James-Large, Withington, TonyBagnall"]
+
+from sktime.networks.base import BaseDeepNetwork
+from sktime.utils.validation._dependencies import _check_soft_dependencies
+
+_check_soft_dependencies("tensorflow", severity="error")
```
Comment (fkiraly) [inline sktime/networks/cnn.py:9]: this should be a warning, and inside the `__init__` with an error.
See the "soft dependencies" docs.

Also, it should be `_check_dl_dependencies`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks ok, but you need to adhere to the dependency handling via checks, see docs for that.

**Comment by TonyBagnall** [comment]
> Looks ok, but you need to adhere to the dependency handling via checks, see docs for that.

tried _check_dl_dependencies but it doesn't have a severity argument, so reverted back to _check_soft_dependencies. I'm not sure why there is a separate method for dl dependencies, will surely just be duplication of code, and I know how much you hate that

**Comment by fkiraly** [comment]
>tried `_check_dl_dependencies` but it doesn't have a severity argument

now it does: https://github.com/alan-turing-institute/sktime/pull/2627

> so reverted back to `_check_soft_dependencies`.

would in-principle be fine by me

> I'm not sure why there is a separate method for dl dependencies, 

more informative (deep learning specific) error message and developer comfort for an anticipated frequent developer case.

> will surely just be duplication of code, and I know how much you hate tha

Agree, could easily be refactored.
Although in this case I would put user and developer comfort above "neat code".

**Comment by TonyBagnall** [comment]
@fkiraly I believe you have someone working on the porting of sktime-dl, which is great. I would like to get this PR in first in order to fix the structure. I've lost track of what you wanted changing here, are you still blocking it, and if so, what do you want changing?

**Comment by fkiraly** [comment]
@TonyBagnall, what have you done/decided about the dependency handling?

Also, it's not "me having someone work" on DL. `sktime`, as a programme & community, is hosting summer projects, one of which is working on this, yes. You are more than welcome to engage with the summer mentees, the meetings are on our calendar, and as CC member you also have a full view of the process (see "updates on GSoC" items).

**Comment by TonyBagnall** [comment]
> @TonyBagnall, what have you done/decided about the dependency handling?
> 
> Also, it's not "me having someone work" on DL. `sktime`, as a programme & community, is hosting summer projects, one of which is working on this, yes. You are more than welcome to engage with the summer mentees, the meetings are on our calendar, and as CC member you also have a full view of the process (see "updates on GSoC" items).

ok, soz my wording offends you, not sure what you mean by "decided about the dependency handling", could you clarify?

**Comment by fkiraly** [comment]
> ok, soz my wording offends you, 

I'm not offended, I just wanted to disagree with the (possibly unintentionally) possessive framing.

> not sure what you mean by "decided about the dependency handling", could you clarify?

Refers to my change request, in which I said:

> Looks ok, but you need to adhere to the dependency handling via checks, see docs for that.

More precisely, I was asking above, for:

* adding informative warning and error messages for developers using the DL functionality
* removing any errors that would be triggered by module import

and with my more recent question, I have asked you for explaining whether and how this has been addressed.

Addressing could mean (a) agreeing with my request and taking some action, or (b) disagreeing and providing a statement why you think you do not need to do it, or (c) thinking something different needs to be done, stating the alternative/refined need and explaining the action taken.

**Comment by TonyBagnall** [comment]
I have switched to your _check_dl_dependencies given you went to the trouble of updating it. I think it already complied with the docs on new soft dependencies (warn/error, imports in functions)

**Comment by TonyBagnall** [comment]
@fkiraly  I would like this in before the next release, could you unblock or give specific changes you would like?

**Comment by xxl4tomxu98** [comment]
is this resolved? e.g. am I able to install the sktime[all_extra] and then all soft dependency are satisfied? including tensorflow?

**Comment by xxl4tomxu98** [comment]
The problem with sktime-dl is that I am using ARM based Mac
 and it won't work when I try to do "conda install sktime-dl"
